oac_var_scope.m4

```diff
@@ -0,0 +1,101 @@
+dnl -*- autoconf -*-
+dnl
+dnl Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
+dnl                         University Research and Technology
+dnl                         Corporation.  All rights reserved.
+dnl Copyright (c) 2004-2018 The University of Tennessee and The University
+dnl                         of Tennessee Research Foundation.  All rights
+dnl                         reserved.
+dnl Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+dnl                         University of Stuttgart.  All rights reserved.
+dnl Copyright (c) 2004-2005 The Regents of the University of California.
+dnl                         All rights reserved.
+dnl Copyright (c) 2007      Sun Microsystems, Inc.  All rights reserved.
+dnl Copyright (c) 2009      Oak Ridge National Labs.  All rights reserved.
+dnl Copyright (c) 2009-2020 Cisco Systems, Inc.  All rights reserved.
+dnl Copyright (c) 2014      Intel, Inc. All rights reserved.
+dnl Copyright (c) 2015-2017 Research Organization for Information Science
+dnl                         and Technology (RIST). All rights reserved.
+dnl Copyright (c) 2021-2022 Amazon.com, Inc. or its affiliates.  All Rights reserved.
+dnl
+dnl $COPYRIGHT$
+dnl
+dnl Additional copyrights may follow
+dnl
+dnl $HEADER$
+
+
+dnl OAC_VAR_SCOPE_INIT
+dnl
+dnl Initialization macro (that is AC_REQUIREd by OAC_VAR_SCOPE_PUSH /
+dnl OAC_VAR_SCOPE_POP) for the var scope subsystem.  Defines the two
+dnl shell functions that implement the configure-time part of the var
+dnl scope evaluation system.
+AC_DEFUN([OAC_VAR_SCOPE_INIT],
```

Comment:
If we ultimately end up with a few things that are converted to shell functions (e.g., the `UNIQ` macro might also be a good target for this), should we have a general `OAC_INIT` macro that simply turns around and invokes `OAC_VAR_SCOPE_INIT` and any others?  This might be a good abstraction so that OMPI/etc. don't have to know all the individual OAC `INIT` macros that need to be invoked.

Or should we ditch the idea of `INIT` macros and hide this kind of stuff behind `AC_REQUIRE`...?  That would be 3% nicer than a top-level `OAC_INIT` macro, because if someone doesn't use `OAC_UNIQ`, then their emitted `configure` doesn't end up including the corresponding shell function for it.  ...but  I remember you said there were some restrictions / rules about how `AC_REQUIRE`s are evaluated, so perhaps there's a reason you didn't do that initially?

Reply:
The `AC_REQUIRE` rules are pretty easy; there's some unexpected dependency with hoisting, and then the issue that we ran into with PMIX (where the expansion of the required macro occurs outside of tree started by configure.m4 and oddness results), but that last one is really a bug in the user code anyway.

I debate on the `OAC_INIT`.  I would not change where we expand the push/pop functions into an `OAC_INIT`, because there's no point in dumping a bunch of shell macros you don't use into configure.  But there's one feature I'd like to enable, which is setting `m4_pattern_forbid([^_?OAC_])` so that there is an error at autoconf time for any unexpanded macros.  We could AC_REQUIRE([_OAC_INIT]) in every OAC macro, or we could require the user to expand `OAC_INIT`.  I'm still debating.  But either way, it would not change this macro, in my opinion.